### PR TITLE
Accomodating PR 23 from `android_device_sony_lilac`: `lineage-17.1` branch

### PR DESCRIPTION
**Pull request by antoniu200**

### CODE REVIEW

File: hardware/OpenCustomizationSelector/src/com/sonymobile/customizationselector/Configurator.java [hunk 1]
```diff
@@ -119,8 +119,13 @@ public boolean isNewConfigurationNeeded() {
...
             }
-            // Actual: (this.mConfigId == null && TextUtils.isEmpty(this.mModem)) ? false : true
-            return mConfigId != null || !TextUtils.isEmpty(mModem);
+            try {
+                return (mConfigId != null || !TextUtils.isEmpty(mModem)) && !ModemSwitcher.getCurrentModemConfig().equals(mModem);
+            } catch (IOException e) {
+                CSLog.w(TAG, "Unable to read out current configuration");
+            	// Actual: (this.mConfigId == null && TextUtils.isEmpty(this.mModem)) ? false : true
+				return mConfigId != null || !TextUtils.isEmpty(mModem);
+            }
```
Comment (Flamefire) [inline hardware/OpenCustomizationSelector/src/com/sonymobile/customizationselector/Configurator.java:128]: I think this part misses actually required updates:

- Note that if `isNewConfigurationNeeded` yields True then `set` will be called setting the config ID **and** modem.
- `mModem` is only non-Empty if it is not equal to `ModemSwitcher.getCurrentModemConfig()`, see `evaluateModem`
- So this means this condition can be rewritten/simplified to `(mConfigId != null && !ModemSwitcher.getCurrentModemConfig().equals(mModem)) || !TextUtils.isEmpty(mModem)) ;`

Given the logic in  `evaluateModem` already taking into account if any changes should be done I assume the same for `evaluateCarrierConfigId` and indeed it checks against `SystemProperties.get(PROP_TA_AC_VERSION, null)`. However that is never set although it should.

So indeed it updates (and reboots) even when it doesn't need to. That is actually a great find from you! I'll fix this together with some bits of other suspicious logic which I think is that way it is due to incomplete decompilation of the original source

File: hardware/OpenCustomizationSelector/src/com/sonymobile/customizationselector/Parser/ModemConfParser.java [hunk 2]
```diff
@@ -75,6 +75,9 @@ private static String parseModemFileName(File modemFile) {
                 CSLog.e(TAG, "IOException: while closing reader", e);
             }
         }
+        if (modemVariant == "" || modemVariant == null) {
```
Comment (Flamefire) [inline hardware/OpenCustomizationSelector/src/com/sonymobile/customizationselector/Parser/ModemConfParser.java:78]: `modemVariant` cannot be `null`. And you also miss the early return at https://github.com/Flamefire/android_device_sony_yoshino-common/pull/40/files#diff-3b685ccc21f9b13cb490b5f12ae11083b5f8c70e04e08b806e1924d951c0c1edR53

Especially the latter point suggests that this isn't the right place to put this code. I mean `parseModemFileName` should just do that (parsing the modem file). The defaulting to another modem should be done higher up. A better place would be `getCarrierBundle` or even at the place of use in `evaluateModem`: Change the parameter value to "ir51_ir92_ims" if it is empty

BTW: Earlier you described this as
> CustomizationSelector was also modified to default to ir51_ir92_ims when IMS Features are enabled and when no customizations are found.

However I don't see an (explicit) check for "when IMS Features are enabled". Could this change use this modem even when IMS is disabled? Or how is this avoided?

Comment (antoniu200) [reply hardware/OpenCustomizationSelector/src/com/sonymobile/customizationselector/Parser/ModemConfParser.java:78]: Okay, yeah, the early return was a significant miss. However, isn't `evaluateModem` executed after `parseModemFileName`?

I also avoided modifying overly abstract functions, so I don't end up breaking something, ~~hence my modification here~~.

As for "when IMS Features are enabled", the parser is turned off when IMS is turned off. Check `EventReceiver.java`: "IMS disabled, not parsing".

EDIT: Also, not sure if other swticher functions are turned off as well when IMS is off, and **this** is the real reason why I made my modification here.

Comment (Flamefire) [reply hardware/OpenCustomizationSelector/src/com/sonymobile/customizationselector/Parser/ModemConfParser.java:78]: > Okay, yeah, the early return was a significant miss. However, isn't `evaluateModem` executed after `parseModemFileName`?

Yep and that's exactly the point: The result of `parseModemFileName` (through any return including the early one) is ultimately passed to `evaluateModem` which has the job of determining the modem to use. So not only do we catch the early return value "automatically" there it also makes more sense to apply the default in a function whose purpose is to evaluate the value/modem-to-use.

> As for "when IMS Features are enabled", the parser is turned off when IMS is turned off. Check `EventReceiver.java`: "IMS disabled, not parsing".

I see. It returns there and doesn't start the `CustomizationSelectorService`

> EDIT: Also, not sure if other swticher functions are turned off as well when IMS is off, and **this** is the real reason why I made my modification here.

If `evaluateModem` is still called we'd need to check that thoroughly if it even makes sense. Or maybe play it safe and check for `(Settings.System.getInt(context.getContentResolver(), "cs_ims", 1) == 0)` before setting the IMS modem as default.

**Comment by antoniu200** [comment]
Sorry to bother here, but do you by any chance remember which camera libraries were incompatible with Android 11 and higher?

Since I only have one G8441 and I still use it as a main device, I would refrain, at least for now, from updating to Android 11. But I would like to see if I can do something with those libraries, most likely try to reverse-engineer them, see how far I can get.

**Comment by Flamefire** [comment]
> Sorry to bother here, but do you by any chance remember which camera libraries were incompatible with Android 11 and higher?

Please do not discuss things in a PR that are unrelated to it as that makes it confusing to understand the PR and actual discussions related to that PR. So just a short answer here: derfelot wrote

> incompatibility of stock library (libimageprocessorjni.so) with new libgui

The former uses some symbols of the latter which are no longer present in A11 which then fails with a linker error at runtime. I spent a considerable amount of time trying to use an old libgui for that library but IIRC it then failed at some other places trying to query camera properties which are not there yet/anymore. But can't remember which ones that was.    
For the first part: LOS uses "shims" to add missing symbols to older stock libraries. That would be the better attempt these days. But that second part will likely make this still impossible. FWIW: In my builds I added a "hacked" Sony camera app which partially works on A11 although with likely restricted features: https://github.com/Flamefire/android_device_sony_lilac/commit/b973a6b78eaa7223f7d20e425ab62c4d02656939

File: hardware/OpenCustomizationSelector/src/com/sonymobile/customizationselector/Configurator.java (hunk 1 above)
Comment (Flamefire) [reply hardware/OpenCustomizationSelector/src/com/sonymobile/customizationselector/Configurator.java:128]: @antoniu200 Closing in on the IMS change and am wondering: Have you tested this part of the change and has it the effect you wanted? Because I can't see how. Thinking about this I believe the change here is a no-op:

You basically added `&& !ModemSwitcher.getCurrentModemConfig().equals(mModem)` to the current condition.

- `mModem` is set to the result of `evaluateModem`/`getModemConfigurationNeeded`
- The only codepath that returns a non-empty string is returning [`getModemConfiguration(variant)`](https://github.com/Flamefire/android_device_sony_yoshino-common/blob/d97d29f4e207026c8a0fe87b91f7684ae0291d11/hardware/OpenCustomizationSelector/src/com/sonymobile/customizationselector/ModemConfiguration.java#L55)
- However it only returns that if `modemConfiguration.equalsIgnoreCase(currentModemConfig)

Hence `mModem` can only be non-empty when it isn't equal to `ModemSwitcher.getCurrentModemConfig()`   
`ModemSwitcher.getCurrentModemConfig()` is never empty (unless an error occured) so an empty `mModem` is different.

Hence `!ModemSwitcher.getCurrentModemConfig().equals(mModem)` is always true and the condition has no effect.

Comment (antoniu200) [reply hardware/OpenCustomizationSelector/src/com/sonymobile/customizationselector/Configurator.java:128]: Yes, I have tested this. The change actually happened because, after I set `ir51_ir92` as default, I was prompted to reboot when swapping SIMs that have no particular modem configuration. The added check seemed to fix the issue.

I was assuming mModem was the new modem to be applied, at least that's what I understood from reading CustomizationSelector's logs.

But I will take a look on my PC in a minute, so I can read the code better.

Comment (antoniu200) [reply hardware/OpenCustomizationSelector/src/com/sonymobile/customizationselector/Configurator.java:128]: Oh, wow. I was staring at the reason why that added condition fixed the reboot issue for almost 20 minutes, trying to explain below why I don't understand what is going on...
When the whole reason was right under my eyes:

`isNewConfigurationNeeded()` originally returned at line 123 a boolean value based on a logic of the sorts:
Is a new Customization Config needed? If yes, we need to reboot.
Is a new Modem needed? If yes, we need to reboot.

So, basically, the reason my phone was rebooting is ~most likely because I still had the original CarrierConfigs, with one for DIGI added.~ because I added a DIGI Customization with the same modem as the one for Orange Romania.
So, I was swapping two SIMs that needed no Modem update, but needed a Customization ID update.

~This is probably why my Sony lost IMS services on DIGI after a while: after upgrading and downgrading between my own builds, the CarrierConfig probably got stuck on the Orange one, hence it only works on Orange now.~

~**That added condition is not needed anymore.**~

It actually still is needed, but can be reduced to `TextUtils.isEmpty(mModem)`.

Given that you have done the refactoring, should I update my PR, or just leave it as it is now?

Comment (Flamefire) [reply hardware/OpenCustomizationSelector/src/com/sonymobile/customizationselector/Configurator.java:128]: > Given that you have done the refactoring, should I update my PR, or just leave it as it is now?

I moved the actual change into a suitable place ([as discussed](https://github.com/Flamefire/android_device_sony_yoshino-common/pull/40#discussion_r1493824251)) with the additional check for IMS to be save, see [this](https://github.com/Flamefire/android_device_sony_yoshino-common/pull/40#discussion_r1494129810) and opened #42 mentioning your handle in the commit. So you don't need to do anything. If you care about commit authorship feel free to commit that yourself and open a PR.

I'll double check the other commits and will merge them soon-ish. https://github.com/antoniu200/android_packages_apps_Settings/commit/907d5897ed90bf1e210c6c42bd958eac388e6c5d looks fine, https://github.com/antoniu200/android_frameworks_opt_telephony/commit/959dcadfec2f7a44f8d884c28f9fa5321ddf6ee0 needs a small change:

```
        boolean provisioned = ImsManager.getInstance(mPhone.getContext(), mPhone.getPhoneId()).isVolteProvisionedOnDevice();
        if (!provisioned && isVolteEnabled) { // Not provisioned, but registered? Carrier probably needs no provisioning.
            ImsManager.getInstance(mPhone.getContext(), mPhone.getPhoneId()).setVolteProvisioned(true);
        }
```

-> 

```
        if (isVolteEnabled) {
            ImsManager imsManager = getImsManager();
            if (!imsManager.isVolteProvisionedOnDevice()) // Not provisioned, but registered? Carrier probably needs no provisioning.
                imsManager.setVolteProvisioned(true); 
        }
```

Basically:
- `ImsManager.getInstance` is used 6 times -> Add a new private function
- No need to check for provisioning if it isn't enabled
- Reuse the `ImsManager` used in the same function

Same for the other 2 changes. And maybe adjust the commit message when you amend&force-push that (would be easiest to create a patch out of it) as currently:
> Otherwise, we end up either enabling more features than available, either none.

", either" should have been likely simply "or". But with the actual change and its comment I'd guess the commit message in total could simply say: "Set provision status for carriers that do not need provisioning", also no need to mention the change file in the commit title. A prefix of e.g. "IMS: " might be better, i.e. the (sub-)system that is changed (like here the commits have a prefix like "OpenCS: " or "SEPolicy: " (if at all)

Comment (antoniu200) [reply hardware/OpenCustomizationSelector/src/com/sonymobile/customizationselector/Configurator.java:128]: I modified my above message a thousandth time, that part of the code drove me nuts: very poor variable naming.
The modification would be nice to have, please check the message again: it actually is very much possible to reboot for nothing.

As for the bug I had with my Sony last time, I fixed it. It was because my provisioning mod required the user to toggle the tech on inside Settings, but that toggle would not be available if the tech was not provisioned: vicious circle.

Hence, the solution was to enable the techs by default.

> I'd guess the commit message in total could simply say: "Set provision status for carriers that do not need provisioning"

Yeah, you could do that, but I chose to write what I wrote as I wanted to explain exactly why I made the changes.

Thing is, Android allows us to disable provisioning when it's not available. Problem is, as mentioned before, we cannot make a million CarrierConfigs, we live in a world where everything should work properly automatically. Also, had I chosen to disable provisioning universally, carriers that needed provisioning would still not have IMS services.

And these features can work automatically very nicely, even when a provisioning server is missing: just check if the phone can register itself in the IMS server for that specific tech. If yes, it means the tech exists and provisioning for it can be set true.

Had I chosen to not do that modification, I risked not enabling IMS techs for carriers that do not need provisioning, as you say, or risked enabling all techs all the time, while some may be missing on the carrier, or on the SIM, etc, which would cause busy tones and calls not going through, hence my "end of the world" joke.

File: hardware/XperiaParts/res/values/strings.xml [hunk 3]
```diff
@@ -38,7 +38,7 @@
     <string name="sim_slot">Sim Slot</string>
     <string name="sim_slot_summary">Select the sim slot you want network switcher to work for.\nSelected Slot: </string>
     <string name="ims_title">IMS Features</string>
-    <string name="ims_summary">Enables IMS features, such as VoLTE, VoWiFi and WiFi calling, <b>if possible</b>.</string>
+    <string name="ims_summary">Enables IMS features, such as VoLTE and VoWiFi, <b>if possible</b>.</string>
```
Comment (antoniu200) [inline hardware/XperiaParts/res/values/strings.xml]: Updated `strings.xml` yet again, to remove duplicate mention of the same tech.

Also worth noting for these changes is that I think I have seen some commit that disabled Video Calling through carrier, not sure why it was done. I can't seem to find it anymore, but am currently looking for it.

Comment (antoniu200) [reply hardware/XperiaParts/res/values/strings.xml]: I can't seem to find that commit anymore, I believe it was from derfelot, although I am not 100% sure.
GitHub's website search also seems broken, at least for me.

**EDIT:** I just found it. It's in lineage-19.1, commit [333a400](https://github.com/Flamefire/android_device_sony_yoshino-common/commit/333a4001fd482f8d13df8b53619a5cdd9c82986d).

File: hardware/OpenCustomizationSelector/src/com/sonymobile/customizationselector/Configurator.java (hunk 1 above)
Comment (Flamefire) [reply hardware/OpenCustomizationSelector/src/com/sonymobile/customizationselector/Configurator.java:128]: > I modified my above message a thousandth time, that part of the code drove me nuts: very poor variable naming. The modification would be nice to have, please check the message again: it actually is very much possible to reboot for nothing.

I checked the message again and you seemingly added 

> So, I was swapping two SIMs that needed no Modem update, but needed a Customization ID update.
> [That added condition] actually still is needed, but can be reduced to TextUtils.isEmpty(mModem).

Which I don't understand. So you do need a Customization ID update but at the same time say you don't need to check for and do that?

To me that first sentence states that the condition should be left as it is now. With the refactoring done it should be a bit easier to understand

Comment (antoniu200) [reply hardware/OpenCustomizationSelector/src/com/sonymobile/customizationselector/Configurator.java:128]: Understandable, let me rephrase.

The condition I added fixed a reboot issue I had that was caused by me adding a DIGI Romania modem customization that was identical to the Orange Romania one.

Currently, `isNewConfigurationNeeded()` returns true if a customization change is needed (regardless of modem) **or** if a modem change is needed.

So, basically, we can currently reboot just because a line residing in file inside `/data/data` (forgot the exact path of OpenCS) was updated.

This update can (and most likely does) happen without rebooting, only the modem update requires a reboot.

Hence, that return value still needs a change, and the best change is to only check if `mModem` is not empty.

**EDIT:** My God, all these changes and comments I made yesterday were probably not even visible to you, because they were "Pending". Great.

Well, I hope you can see them now.

Comment (Flamefire) [reply hardware/OpenCustomizationSelector/src/com/sonymobile/customizationselector/Configurator.java:128]: > Currently, `isNewConfigurationNeeded()` returns true if a customization change is needed (regardless of modem) **or** if a modem change is needed.
> 
> So, basically, we can currently reboot just because a line residing in file inside `/data/data` (forgot the exact path of OpenCS) was updated.

But we need to update the customization (i.e. MiscTA) if a customization change is needed. If only checking for a modem change than we will miss that change which likely will lead to missing features.

From what I understand what happens is: You write to MiscTA. Upon reboot the hardware/driver will write into a property. This property is then checked against the desired value in `evaluateConfig` which will return `null` if no change is required which makes this function return False, hence no update and no reboot will happen.   
However I haven't found what/where exactly that property is written, so I just assume it is on reboot.

I fixed a potential issue where a change to that config was detected even if it isn't required: https://github.com/Flamefire/android_device_sony_yoshino-common/commit/342abc1c9e22650ae5d3b5690f843a0c4dbef843

I also improved the logging to provide the value in the property and the MiscTA to aid in debugging: https://github.com/Flamefire/android_device_sony_yoshino-common/blob/dc325bcf9b5d720e0a8fc8ced82b8126ded6b597/hardware/OpenCustomizationSelector/src/com/sonymobile/customizationselector/Configurator.java#L84-L86

> This update can (and most likely does) happen without rebooting, only the modem update requires a reboot.

As mentioned I'm not sure about that so I'd rather reboot one more time than one to less. I mean it only happens when changing SIMs, doesn't it? That's rare enough. Furthermore at this point you are actually **preventing** the change in the first place, so with only checking the modem here the config will **not** be updated if it would only be the config in need of updating. This is much more likely after defaulting to the ir-modem for IMS.

> **EDIT:** My God, all these changes and comments I made yesterday were probably not even visible to you, because they were "Pending". Great.
> 
> Well, I hope you can see them now.

You mean the ones on `strings.xml`? If so yes.

Comment (antoniu200) [reply hardware/OpenCustomizationSelector/src/com/sonymobile/customizationselector/Configurator.java:128]: > But we need to update the customization (i.e. MiscTA) if a customization change is needed. If only checking for a modem change than we will miss that change which likely will lead to missing features.

Oh, no, it is not the case. We are not changing CarrierConfigs using OpenCS, but rather only modems, which are specified using Customization IDs (`mConfigID` in the code above).

> From what I understand what happens is: You write to MiscTA.

I have to check this, but, AFAIK, MiscTA is in /oem somewhere, however my /oem is completely empty and OpenCS works fine.

> I fixed a potential issue where a change to that config was detected even if it isn't required: [342abc1](https://github.com/Flamefire/android_device_sony_yoshino-common/commit/342abc1c9e22650ae5d3b5690f843a0c4dbef843)
> 
> I also improved the logging to provide the value in the property and the MiscTA to aid in debugging:
> 
> https://github.com/Flamefire/android_device_sony_yoshino-common/blob/dc325bcf9b5d720e0a8fc8ced82b8126ded6b597/hardware/OpenCustomizationSelector/src/com/sonymobile/customizationselector/Configurator.java#L84-L86

Will check these a bit later on my PC as well.

> Furthermore at this point you are actually **preventing** the change in the first place, so with only checking the modem here the config will **not** be updated if it would only be the config in need of updating. This is much more likely after defaulting to the ir-modem for IMS.

The point is definitely not to prevent the change, but rather just the reboot.

However, I am not sure I understand what you mean by "config" here. OpenCS' config or MiscTA? Since the first one is definitely updated at runtime.

> > **EDIT:** My God, all these changes and comments I made yesterday were probably not even visible to you, because they were "Pending". Great.
> > Well, I hope you can see them now.
> 
> You mean the ones on `strings.xml`? If so yes.

Good. Even my previous comment was pending for a while before I realised what I had to do to publish it.

Comment (Flamefire) [reply hardware/OpenCustomizationSelector/src/com/sonymobile/customizationselector/Configurator.java:128]: > > But we need to update the customization (i.e. MiscTA) if a customization change is needed. If only checking for a modem change than we will miss that change which likely will lead to missing features.
> 
> Oh, no, it is not the case. We are not changing CarrierConfigs using OpenCS, but rather only modems, which are specified using Customization IDs (`mConfigID` in the code above).

There is `mConfigID` and `mModem`. Both are written to different parts in the `MiscTA` "partition" using the class of the same name.

> The point is definitely not to prevent the change, but rather just the reboot.

Then this approach does not work as this function is used by `set` to determine if anything needs setting and by callers to determine if `set` needs to be called.

> However, I am not sure I understand what you mean by "config" here. OpenCS' config or MiscTA? Since the first one is definitely updated at runtime.

`mConfig` basically.

Edit: I just verified that my analysis is correct: When `Configurator` writes to MiscTA using `MiscTA.write(TA_AC_VERSION, mConfigId)` the value of the property `PROP_TA_AC_VERSION` is not updated until a reboot. Hence it makes sense to write `mConfigId` and (potentially) reboot when it is different to `PROP_TA_AC_VERSION` (i.e. `evaluateCarrierConfigId` returns a non-null value)

File: overlay/frameworks/base/core/res/res/values/config.xml [hunk 4]
```diff
@@ -264,7 +264,7 @@
     <bool name="config_cellBroadcastAppLinks">true</bool>
 
     <!-- Is the device capable of hot swapping an ICC Card -->
-    <bool name="config_hotswapCapable">true</bool>
+    <bool name="config_hotswapCapable">false</bool>
```
Comment (Flamefire) [inline overlay/frameworks/base/core/res/res/values/config.xml:267]: I'm pretty sure this change is not needed. We explicitely handle the `ACTION_CARRIER_CONFIG_CHANGED` event which triggers `evaluateCarrierBundle` that prompts for a reboot if required. With this change Android will always prompt for a reboot

Comment (antoniu200) [reply overlay/frameworks/base/core/res/res/values/config.xml:267]: There are two problems with triggering `ACTION_CARRIER_CONFIG_CHANGED`:

1. After a SIM is removed, the modem dies completely, so there will be no more SIM to evaluate a carrier for, the device must be rebooted.
2. Even if the modem didn't die completely, we now have only one Carrier Config applied for **a bunch** of operators.

Practice tells a similar story, trying to swap a SIM currently leads to a dead modem until reboot, with absolutely no reboot dialog ever appearing, even after inserting any SIM back in.

Comment (Flamefire) [reply overlay/frameworks/base/core/res/res/values/config.xml:267]: > After a SIM is removed, the modem dies completely, so there will be no more SIM to evaluate a carrier for, the device must be rebooted.

I just tried and you are right. Didn't expect that but that's how it is. So this change is ok then.

> Even if the modem didn't die completely, we now have only one Carrier Config applied for a bunch of operators.

I don't understand that. If we need a different Carrier Config it will/would be applied and prompt for a reboot. Doesn't matter though as due to the modem bug above we can't hotswap...

Comment (antoniu200) [reply overlay/frameworks/base/core/res/res/values/config.xml:267]: > I don't understand that. If we need a different Carrier Config it will/would be applied and prompt for a reboot. Doesn't matter though as due to the modem bug above we can't hotswap...

I went a bit off-topic with the explanation.

The point was that not all SIM hotswaps require a CarrerConfig change, since not all operators have a CarrierConfig on this phone, so, even if the modem didn't die completely, the phone would still not reboot automatically.

File: hardware/OpenCustomizationSelector/src/com/sonymobile/customizationselector/Configurator.java (hunk 1 above)
Comment (antoniu200) [reply hardware/OpenCustomizationSelector/src/com/sonymobile/customizationselector/Configurator.java:128]: What I notice is that writing to MiscTA is done using a function called `writeRAW()`, but I cannot find it anywhere in the repo. So, does that function work? If yes, where is it?

Comment (Flamefire) [reply hardware/OpenCustomizationSelector/src/com/sonymobile/customizationselector/Configurator.java:128]: `writeRaw` is declared as "native", hence it lives inside a native binary, not in Java. In this case you can even see which one easily as it is mentioned right next to it: https://github.com/Flamefire/android_device_sony_yoshino-common/blob/9202c964dd848cf507530f4e7483eb74495ea1ce/hardware/OpenCustomizationSelector/src-misc-ta/com/sonymobile/miscta/MiscTA.java#L14

See https://github.com/Flamefire/android_device_sony_lilac/blob/1695c51519186d66928439ca5ed65c7263380866/proprietary-files.txt#L93

Comment (antoniu200) [reply hardware/OpenCustomizationSelector/src/com/sonymobile/customizationselector/Configurator.java:128]: Okay, I do not like the current solution, but it seems we have to leave it like this, as of now. I cannot tell what `writeRAW` from inside `libMiscTAApi.sony.so` does, as I barely have any experience with ARMv7 assembly and, besides, ARM decompilers can barely interpret the results of C code intended for hardware / drivers: many missing external functions, function calls, wrong and unfixable function arguments, etc.

**Comment by antoniu200** [comment]
I think everything here has been commited. I am closing this PR now.

**Comment by Flamefire** [comment]
Right, I cherry-picked the commits for conflict resolution and testing. Seems to work. Thanks!
